src/api/app/models/owner_search.rb

```diff
@@ -0,0 +1,134 @@
+require 'api_exception'
+
+class OwnerSearch
+  class AttributeNotSetError < APIException
+    setup 'attribute_not_set', 400
+  end
+
+  def initialize(params = {})
+    self.params = params
+    self.attribute = AttribType.find_by_name!(params[:attribute] || 'OBS:OwnerRootProject')
+
+    self.limit = (params[:limit] || 1).to_i
+  end
+
+  def for(owner)
+    # search in each marked project
+    object_projects(owner).map do |project|
+      find_containers(project, owner, !devel_disabled?(project))
+    end.flatten
+  end
+
+  protected
+
+  attr_accessor :params, :attribute, :limit
+
+  def object_projects(_obj)
+    # default project specified
+    return [Project.get_by_name(params[:project])] if params[:project]
+
+    # Find all marked projects
+    projects = Project.find_by_attribute_type(attribute)
+    return projects unless projects.empty?
+    raise AttributeNotSetError, "The attribute #{attribute.fullname} is not set to define default projects."
+  end
+
+  def project_attrib(project)
+    return unless project
+    project.attribs.find_by(attrib_type: attribute)
+  end
+
+  def filter(project)
+    return params[:filter].split(',') if params[:filter]
+
+    attrib = project_attrib(project)
+    if attrib && attrib.values.where(value: 'BugownerOnly').exists?
+      ['bugowner']
+    else
+      ['maintainer', 'bugowner']
+    end
+  end
+
+  def devel_disabled?(project = nil)
+    return ['0', 'false'].include? params[:devel] if params[:devel]
+
+    attrib = project_attrib(project)
+    attrib && attrib.values.where(value: 'DisableDevel').exists?
+  end
+
+  def find_containers(rootproject, owner, devel)
+    projects = rootproject.expand_all_projects(allow_remote_projects: false)
+
+    roles = []
+    filter(rootproject).each do |f|
+      roles << Role.find_by_title!(f)
+    end
+
+    found_packages = Relationship.where(role_id: roles, package_id: Package.where(project_id: projects).pluck(:id))
+    found_projects = Relationship.where(role_id: roles, project_id: projects)
+    # fast find packages with defintions
+    if owner.class == User
+      # user in package object
+      found_packages = found_packages.where(user_id: owner)
+      # user in project object
+      found_projects = found_projects.where(user_id: owner)
+    elsif owner.class == Group
+      # group in package object
+      found_packages = found_packages.where(group_id: owner)
+      # group in project object
+      found_projects = found_projects.where(group_id: owner)
+    else
```

Comment:
How about...

```ruby
if owner.class == User
  query = user_id: owner
elsif owner.class == Group
  query = group_id: owner
else
  raise
end

found_packages = found_packages.where(query)
found_projects = found_projects.where(query)
```
...?

Reply:
much better

```
query = case owner
when User
 user_id: owner
when Group:
 group_id: owner
else
  raise
end

```

Reply:
I'm in the process of refactoring this method also - I ignored it so far

Reply:
Check 9aa8feac4cbbc5a2a406b0935080d6d095145222